File: gc.c
#include <stdio.h>
#include <stdlib.h>
#include <stdint.h>

typedef uint64_t SNAKEVAL;

void printHelp(FILE* out, SNAKEVAL val);
extern uint64_t NUM_TAG_MASK;
extern uint64_t CLOSURE_TAG_MASK;
extern uint64_t TUPLE_TAG_MASK;
extern uint64_t CLOSURE_TAG;
extern uint64_t TUPLE_TAG;
extern uint64_t NIL;
extern uint64_t tupleCounter;
extern uint64_t* STACK_BOTTOM;
extern uint64_t* FROM_S;
extern uint64_t* FROM_E;
extern uint64_t* TO_S;
extern uint64_t* TO_E;

void naive_print_heap(uint64_t* heap, uint64_t* heap_end) {
  printf("In naive_print_heap from %p to %p\n", heap, heap_end);
  for(uint64_t i = 0; i < (uint64_t)(heap_end - heap); i += 1) {
    printf("  %ld/%p: %p (%ld)\n", i, (heap + i), (uint64_t*)(*(heap + i)), *(heap + i));
  }
}

// Implement the functions below

void smarter_print_heap(uint64_t* from_start, uint64_t* from_end, uint64_t* to_start, uint64_t* to_end) {
  // Print out the entire heap (both semispaces), and
  // try to print values readably when possible
}

/*
  Copies a Garter value from the given address to the new heap, 
  but only if the value is heap-allocated and needs copying.

  Arguments:
    garter_val_addr: the *address* of some Garter value, which contains a Garter value,
                     i.e. a tagged word.  
                     It may or may not be a pointer to a heap-allocated value...
    heap_top: the location at which to begin copying, if any copying is needed

  Return value:
    The new top of the heap, at which to continue allocations

  Side effects:
    If the data needed to be copied, then this replaces the value at its old location 
    with a forwarding pointer to its new location
 */
uint64_t* copy_if_needed(uint64_t* garter_val_addr, uint64_t* heap_top) {
  uint64_t val = *garter_val_addr;

  // fprintf(stderr, "  copy_if_needed: addr=%p val=0x%lx tag=%ld\n",
  //         garter_val_addr, val, val & 0x7);

  // Check if tuple (tag 0x1, but not NIL which is exactly 0x1)
  if ((val & 0x7) == TUPLE_TAG && val != NIL) {
    uint64_t* old_addr = (uint64_t*)(val - TUPLE_TAG);
    uint64_t first_word = old_addr[0];

    // Check if already forwarded (odd first word = forwarding pointer)
    if (first_word & 0x1) {
      // Already forwarded: update the reference to point to new location
      *garter_val_addr = (first_word - 1) | TUPLE_TAG;
      return heap_top;
    }

    // Copy tuple: first_word is length*2, so actual length is first_word/2
    uint64_t len = first_word / 2;
    uint64_t words = 1 + len;  // header + elements
    for (uint64_t i = 0; i < words; i++) {
      heap_top[i] = old_addr[i];
    }

    // Install forwarding pointer (new address + 1 to mark as forwarded)
    old_addr[0] = ((uint64_t)heap_top) | 0x1;

    // Update the reference to point to new location with tuple tag
    *garter_val_addr = ((uint64_t)heap_top) | TUPLE_TAG;

    // Advance heap_top (16-byte aligned)
    return heap_top + words + (words % 2);
  }

  // Check if closure (tag 0x5)
  if ((val & 0x7) == CLOSURE_TAG) {
    uint64_t* old_addr = (uint64_t*)(val - CLOSURE_TAG);
    uint64_t first_word = old_addr[0];

    // Check if already forwarded
    if (first_word & 0x1) {
      *garter_val_addr = (first_word - 1) | CLOSURE_TAG;
      return heap_top;
    }

    // Closure layout: [arity<<2, code_ptr, num_frees, free0, free1, ...]
    uint64_t num_frees = old_addr[2];
    uint64_t words = 3 + num_frees;  // arity, code_ptr, num_frees, + free vars
    for (uint64_t i = 0; i < words; i++) {
      heap_top[i] = old_addr[i];
    }

    // Install forwarding pointer
    old_addr[0] = ((uint64_t)heap_top) | 0x1;

    // Update the reference
    *garter_val_addr = ((uint64_t)heap_top) | CLOSURE_TAG;

    // Advance (16-byte aligned)
    return heap_top + words + (words % 2);
  }

  // Not a heap pointer (number, boolean, nil), return unchanged
  return heap_top;
}

/*
  Implements Cheney's garbage collection algorithm.

  Arguments:
    bottom_frame: the base pointer of our_code_starts_here, i.e. the bottommost Garter frame
    top_frame: the base pointer of the topmost Garter stack frame
    top_stack: the current stack pointer of the topmost Garter stack frame
    from_start and from_end: bookend the from-space of memory that is being compacted
    to_start: the beginning of the to-space of memory

  Returns:
    The new location within to_start at which to allocate new data
 */
uint64_t* gc(uint64_t* bottom_frame, uint64_t* top_frame, uint64_t* top_stack, uint64_t* from_start, uint64_t* from_end, uint64_t* to_start) {

  uint64_t* alloc_ptr = to_start;  // Where to allocate next
  uint64_t* scan_ptr = to_start;   // Where to scan next (Cheney's algorithm)

  // fprintf(stderr, "GC: bottom=%p top_frame=%p top_stack=%p from_start=%p to_start=%p\n",
  //         bottom_frame, top_frame, top_stack, from_start, to_start);

  // Phase 1: Copy all roots from the stack
  uint64_t* old_top_frame = top_frame;
  do {
    for (uint64_t* cur_word = top_stack /* maybe need a +1 here? */; cur_word < top_frame; cur_word++) {
      alloc_ptr = copy_if_needed(cur_word, alloc_ptr);
    }
    /* Shift to next stack frame:
     * [top_frame] points to the saved RBP, which is the RBP of the next stack frame,
     * [top_frame + 8] is the return address, and
     * [top_frame + 16] is therefore the next frame's stack-top
     */
    top_stack = top_frame + 2;
    old_top_frame = top_frame;
    top_frame = (uint64_t*)(*top_frame);
  } while (old_top_frame < bottom_frame); // Use the old stack frame to decide if there's more GC'ing to do

  // Phase 2: Cheney scan - scan copied objects and copy their children
  while (scan_ptr < alloc_ptr) {
    uint64_t first_word = scan_ptr[0];

    // Determine object type from header
    // Tuple: first_word = length*2 (even, could be 2, 4, 6, 8, ...)
    // Closure: first_word = arity<<2 (multiple of 4: 0, 4, 8, 12, ...)
    //
    // Distinguishing criterion:
    // - If first_word % 4 == 2, it's definitely a tuple (odd-length tuple)
    // - If first_word % 4 == 0, check second word:
    //   - Closure: second word is a code pointer (8-byte aligned, high address > heap)
    //   - Tuple: second word is a Garter value (will have tag bits or be in heap range)
    //
    // Better heuristic: check if second word is in the from-space or to-space range
    // If not, it's likely a code pointer (closure).

    int is_closure = 0;
    if ((first_word & 0x3) == 0) {
      // Could be closure (arity*4) or even-length tuple (len*2 where len is even)
      uint64_t second_word = scan_ptr[1];

      // Code pointers are not in heap memory
      // If second_word is outside both from-space and to-space, it's a code pointer
      int in_from_space = (second_word >= (uint64_t)FROM_S && second_word < (uint64_t)FROM_E);
      int in_to_space = (second_word >= (uint64_t)TO_S && second_word < (uint64_t)TO_E);

      // Also check: Garter values have tags in low bits, code pointers are 8-byte aligned
      // Code pointers won't have tag patterns (1 for tuple, 5 for closure)
      int looks_like_garter_heap_val = ((second_word & 0x7) == 1 || (second_word & 0x7) == 5);

      // If second_word looks like a code pointer (not in heap, high address)
      if (!in_from_space && !in_to_space && !looks_like_garter_heap_val &&
          second_word > 0x100000000UL) {
        is_closure = 1;
      }
    }

    if (is_closure) {
      // Closure: [arity<<2, code_ptr, num_frees, free0, ...]
      uint64_t num_frees = scan_ptr[2];
      // Copy each free variable (they start at offset 3)
      for (uint64_t i = 0; i < num_frees; i++) {
        alloc_ptr = copy_if_needed(&scan_ptr[3 + i], alloc_ptr);
      }
      uint64_t words = 3 + num_frees;
      scan_ptr += words + (words % 2);  // 16-byte aligned
    } else {
      // Tuple: [length*2, elem0, elem1, ...]
      uint64_t len = first_word / 2;
      // Copy each element
      for (uint64_t i = 0; i < len; i++) {
        alloc_ptr = copy_if_needed(&scan_ptr[1 + i], alloc_ptr);
      }
      uint64_t words = 1 + len;
      scan_ptr += words + (words % 2);  // 16-byte aligned
    }
  }

  // after copying and GC'ing all the stack frames, return the new allocation starting point
  return alloc_ptr;
}

File: runtime/src/gc.rs
//! Cheney's copying garbage collector

use crate::heap::*;
use crate::values::*;
use std::ptr;

const FORWARDING_MARKER: u64 = 0x1;
const STRING_LENGTH_THRESHOLD: u64 = 1_000_000;
const MIN_CODE_POINTER: u64 = 0x1_0000_0000;

#[inline]
fn is_forwarded(first_word: u64) -> bool {
    (first_word & FORWARDING_MARKER) != 0
}

#[inline]
fn forwarded_address(first_word: u64) -> u64 {
    first_word - FORWARDING_MARKER
}

#[inline]
fn is_in_space(addr: u64, start: *mut u64, end: *mut u64) -> bool {
    addr >= start as u64 && addr < end as u64
}

#[inline]
fn looks_like_heap_value(word: u64) -> bool {
    matches!(word & 0x7, 1 | 3 | 5)
}

#[inline]
fn align_words(words: u64) -> u64 {
    words + (words & 1)
}

enum HeapObjectKind {
    Tuple { len: u64 },
    String { char_words: u64 },
    Closure { num_frees: u64 },
}

/// Detect the kind of heap object at the given pointer
///
/// # Safety
/// scan_ptr must point to a valid heap object header
unsafe fn detect_object_kind(scan_ptr: *mut u64, from_start: *mut u64, from_end: *mut u64, to_start: *mut u64, to_end: *mut u64) -> HeapObjectKind {
    let first_word = *scan_ptr;

    // Check if first_word is odd and small (likely a string length)
    if is_forwarded(first_word) && first_word <= STRING_LENGTH_THRESHOLD {
        let char_words = (first_word + 7) / 8;
        return HeapObjectKind::String { char_words };
    }

    // Check if this could be a closure
    if (first_word & 0x3) == 0 {
        let second_word = *scan_ptr.add(1);

        // Code pointers are not in heap memory
        let in_from_space = is_in_space(second_word, from_start, from_end);
        let in_to_space = is_in_space(second_word, to_start, to_end);

        // If second_word looks like a code pointer (not in heap, high address, aligned)
        if !in_from_space
            && !in_to_space
            && !looks_like_heap_value(second_word)
            && second_word > MIN_CODE_POINTER
            && (second_word & 0x7) == 0
        {
            let num_frees = (*scan_ptr.add(1)) / 2;
            return HeapObjectKind::Closure { num_frees };
        }
    }

    HeapObjectKind::Tuple { len: first_word / 2 }
}

unsafe fn copy_object(old_addr: *mut u64, heap_top: *mut u64, words: u64, tag: u64, garter_val_addr: *mut u64) -> *mut u64 {
    ptr::copy_nonoverlapping(old_addr, heap_top, words as usize);

    *old_addr = (heap_top as u64) | FORWARDING_MARKER;

    *garter_val_addr = (heap_top as u64) | tag;

    heap_top.add(align_words(words) as usize)
}

pub unsafe fn copy_if_needed(garter_val_addr: *mut u64, heap_top: *mut u64) -> *mut u64 {
    let val = *garter_val_addr;

    if (val & 0x7) == TUPLE_TAG && val != NIL {
        let old_addr = (val - TUPLE_TAG) as *mut u64;
        let first_word = *old_addr;

        if is_forwarded(first_word) {
            *garter_val_addr = forwarded_address(first_word) | TUPLE_TAG;
            return heap_top;
        }

        let len = first_word / 2;
        let words = 1 + len; // header + elements
        return copy_object(old_addr, heap_top, words, TUPLE_TAG, garter_val_addr);
    }

    if (val & 0x7) == STRING_TAG {
        let old_addr = (val - STRING_TAG) as *mut u64;
        let first_word = *old_addr;

        if is_forwarded(first_word) {
            *garter_val_addr = forwarded_address(first_word) | STRING_TAG;
            return heap_top;
        }

        let len = first_word;
        let char_words = (len + 7) / 8;
        let words = 1 + char_words;
        return copy_object(old_addr, heap_top, words, STRING_TAG, garter_val_addr);
    }

    if (val & 0x7) == CLOSURE_TAG {
        let old_addr = (val - CLOSURE_TAG) as *mut u64;
        let first_word = *old_addr;

        if is_forwarded(first_word) {
            *garter_val_addr = forwarded_address(first_word) | CLOSURE_TAG;
            return heap_top;
        }

        let num_frees = (*old_addr.add(1)) / 2;
        let words = 3 + num_frees;
        return copy_object(old_addr, heap_top, words, CLOSURE_TAG, garter_val_addr);
    }

    // Not a heap pointer (number, boolean, nil), return unchanged
    heap_top
}

pub unsafe fn gc(
    bottom_frame: *mut u64,
    top_frame: *mut u64,
    top_stack: *mut u64,
    _from_start: *mut u64,
    _from_end: *mut u64,
    to_start: *mut u64,
) -> *mut u64 {
    let mut alloc_ptr = to_start;
    let mut scan_ptr = to_start;

    let mut current_top_frame = top_frame;
    let mut current_top_stack = top_stack;
    let mut old_top_frame;

    loop {
        let mut cur_word = current_top_stack;
        while cur_word < current_top_frame {
            alloc_ptr = copy_if_needed(cur_word, alloc_ptr);
            cur_word = cur_word.add(1);
        }

        old_top_frame = current_top_frame;
        current_top_stack = current_top_frame.add(2);
        current_top_frame = *current_top_frame as *mut u64;

        if old_top_frame >= bottom_frame {
            break;
        }
    }

    while scan_ptr < alloc_ptr {
        let kind = detect_object_kind(scan_ptr, FROM_S, FROM_E, TO_S, TO_E);

        match kind {
            HeapObjectKind::String { char_words } => {
                let words = 1 + char_words;
                scan_ptr = scan_ptr.add(align_words(words) as usize);
            }
            HeapObjectKind::Closure { num_frees } => {
                for i in 0..num_frees {
                    alloc_ptr = copy_if_needed(scan_ptr.add(3 + i as usize), alloc_ptr);
                }
                let words = 3 + num_frees;
                scan_ptr = scan_ptr.add(align_words(words) as usize);
            }
            HeapObjectKind::Tuple { len } => {
                for i in 0..len {
                    alloc_ptr = copy_if_needed(scan_ptr.add(1 + i as usize), alloc_ptr);
                }
                let words = 1 + len;
                scan_ptr = scan_ptr.add(align_words(words) as usize);
            }
        }
    }

    alloc_ptr
}

/// Try to reserve memory, running GC if needed.
///
/// # Arguments
/// * `alloc_ptr` - Current heap allocation pointer (R15)
/// * `bytes_needed` - Number of bytes needed
/// * `cur_frame` - Current RBP
/// * `cur_stack_top` - Current RSP
///
/// # Returns
/// New allocation pointer after GC
///
/// # Safety
/// This function performs garbage collection.
pub unsafe fn try_gc(
    _alloc_ptr: *mut u64,
    bytes_needed: u64,
    cur_frame: *mut u64,
    cur_stack_top: *mut u64,
) -> *mut u64 {
    let layout = std::alloc::Layout::from_size_align((HEAP_SIZE + 15) * 8, 8).unwrap();
    let new_heap = std::alloc::alloc_zeroed(layout) as *mut u64;
    let old_heap = HEAP;

    if new_heap.is_null() {
        eprintln!("Out of memory: could not allocate a new semispace for garbage collection");
        if !old_heap.is_null() {
            std::alloc::dealloc(old_heap as *mut u8, layout);
        }
        std::process::exit(crate::errors::ERR_OUT_OF_MEMORY as i32);
    }

    let new_r15 = ((new_heap as u64 + 15) & !0xF) as *mut u64;
    let new_heap_end = new_r15.add(HEAP_SIZE);

    FROM_S = ((HEAP as u64 + 15) & !0xF) as *mut u64;
    FROM_E = HEAP_END;
    TO_S = new_r15;
    TO_E = new_heap_end;

    let new_r15 = gc(STACK_BOTTOM, cur_frame, cur_stack_top, FROM_S, HEAP_END, new_r15);

    HEAP = new_heap;
    HEAP_END = new_heap_end;

    std::alloc::dealloc(old_heap as *mut u8, layout);

    let words_needed = bytes_needed / 8;
    let heap_size = HEAP_SIZE;
    if words_needed > heap_size as u64 {
        eprintln!(
            "Allocation error: needed {} words, but the heap is only {} words",
            words_needed, heap_size
        );
        std::alloc::dealloc(new_heap as *mut u8, layout);
        std::process::exit(crate::errors::ERR_OUT_OF_MEMORY as i32);
    }

    let remaining = (HEAP_END as usize - new_r15 as usize) / 8;
    if words_needed > remaining as u64 {
        eprintln!(
            "Out of memory: needed {} words, but only {} remain after collection",
            words_needed, remaining
        );
        std::alloc::dealloc(new_heap as *mut u8, layout);
        std::process::exit(crate::errors::ERR_OUT_OF_MEMORY as i32);
    }

    new_r15
}

pub unsafe fn naive_print_heap(heap: *mut u64, heap_end: *mut u64) -> *mut u64 {
    println!(
        "In naive_print_heap from {:p} to {:p}",
        heap as *const u64, heap_end as *const u64
    );
    let count = (heap_end as usize - heap as usize) / 8;
    for i in 0..count {
        let addr = heap.add(i);
        let val = *addr;
        println!("  {}/{:p}: {:p} ({})", i, addr, val as *const u64, val);
    }
    heap
}
